ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java

```diff
@@ -369,4 +370,76 @@ private Schema buildAggregateSchema(
+  ) {
+    return expressionList.stream()
+        .map(argExpression -> argExpression instanceof Literal
+                              ? argExpression
+                              : new QualifiedNameReference(
+            QualifiedName.of(
+                internalSchema
+                    .getExpressionToInternalColumnNameMap()
+                    .get(argExpression.toString()))))
+        .collect(Collectors.toList());
+  }
+
+  private List<Pair<String, Expression>> updateFinalSelectExpressions(
+      final List<Pair<String, Expression>> finalSelectExpressions,
+      final Map<String, String> expressionToInternalColumnNameMap) {
+    return finalSelectExpressions.stream()
+        .map(finalSelectExpression ->
+                 expressionToInternalColumnNameMap
+                     .containsKey(finalSelectExpression.getRight().toString())
+                 ? new Pair<>(finalSelectExpression.getLeft(),
+                            (Expression)
+                            new QualifiedNameReference(
+                                QualifiedName.of(expressionToInternalColumnNameMap
+                                          .get(finalSelectExpression.getRight().toString()))))
+                 : new Pair<>(finalSelectExpression.getLeft(), finalSelectExpression.getRight()))
+        .collect(Collectors.toList());
+  }
+
+  class InternalSchema {
+    private final List<Pair<String, Expression>> aggArgExpansionList = new ArrayList<>();
+    private final Map<String, Integer> internalNameToIndexMap = new HashMap<>();
+    private final Map<String, String> expressionToInternalColumnNameMap = new HashMap<>();
+
+    public InternalSchema(
+        final List<Expression> requiredColumnList,
+        final List<Expression> aggregateFunctionArguments) {
+      collectAggregateArgExpressions(requiredColumnList);
+      collectAggregateArgExpressions(aggregateFunctionArguments);
+    }
+
+
+    private void collectAggregateArgExpressions(
+        final List<Expression> expressions
+    ) {
+      expressions.stream()
+          .filter(e -> !internalNameToIndexMap.containsKey(e.toString()))
+          .forEach(expression -> {
+            String internalColumnName = INTERNAL_COLUMN_NAME_PREFIX + aggArgExpansionList.size();
+            internalNameToIndexMap.put(internalColumnName, aggArgExpansionList.size());
+            aggArgExpansionList.add(new Pair<>(internalColumnName, expression));
+            if (!expressionToInternalColumnNameMap.containsKey(expression.toString())) {
+              expressionToInternalColumnNameMap.put(expression.toString(), internalColumnName);
+            }
+          });
+    }
+
+    public List<Pair<String, Expression>> getAggArgExpansionList() {
+      return aggArgExpansionList;
+    }
+
+    public Map<String, Integer> getInternalNameToIndexMap() {
+      return internalNameToIndexMap;
+    }
+
+    public Map<String, String> getExpressionToInternalColumnNameMap() {
```

Comment:
Continuing from my comments elsewhere. If we had higher level methods which were more semantically meaningful, this would not have to be a public function.